File: runtime/resume.go
package runtime

import (
	"regexp"
	"strconv"
	"strings"
)

// bootstrapRuntimeCode is the walker runtime prelude, emitted at most once per
// render. In production Marko this is a hardcoded string constant with the
// runtime and render ids interpolated; it never varies at the pinned version
// (@marko/runtime-tags 6.3.42) except for those two ids.
const bootstrapRuntimeCode = `(e=>(self[e]||=(l,f=e+l,s=f.length,a={},d=[],t=document,n=t.createTreeWalker(t,129))=>t=self[e][l]={i:f,d:t,l:a,v:d,x(){},w(e,l,r){for(;e=n.nextNode();)t.x(l=(l=e.data)&&!l.indexOf(f)&&(a[r=l.slice(s+1)]=e,l[s]),r,e),l>"#"&&d.push(e)}},self[e]))`

// idPattern is the validation Marko's render() applies to runtimeId and
// renderId before interpolating them UNESCAPED into the bootstrap. Go applies
// the same rule so the payload cannot be broken by a bad id.
var idPattern = regexp.MustCompile(`^[_a-zA-Z][_a-zA-Z0-9]*$`)

// Default runtime and render ids. "M" is always the runtime id; "_" is the
// render id for PAGE templates, which is all go-marko renders. (An embedded
// template would get a random render id, losing byte-determinism.)
const (
	DefaultRuntimeID = "M"
	DefaultRenderID  = "_"
)

// Resume op codes. Only OpResume is needed for flat reactive templates; the
// branch ops arrive with <if>/<for> and the reorder ops with streaming.
//
// Resume support; generated code only; not stable.
const (
	// OpResume ("*") binds the preceding sibling node into scope[accessor].
	OpResume = "*"
)

// resumeState is the Writer's resume channel: the scope partials, effects
// string and flags accumulated during markup emission, plus the serializer
// that turns them into payload bytes at flush time.
type resumeState struct {
	runtimeID string
	renderID  string

	// writeScopes accumulates per-scope partials, MERGED rather than
	// replaced when the same id is written more than once.
	writeScopes map[int]*ScopeState

	// nextScopeID is the render's scope-id allocator. 1-based: 0 is $global.
	nextScopeID int

	// effects is the script-entry token stream; lastEffect tracks the
	// current registry id so consecutive entries with the same id collapse.
	effects      string
	lastEffect   string
	hasEffect    bool
	needsRuntime bool

	// clientBundle is the URL of this page's browser bundle, emitted as a
	// module script AFTER the payload script and BEFORE the trailers. See
	// Writer.ClientBundle.
	clientBundle string

	ser *serializer
	err error
}

// ClientBundle names the browser bundle for this page. FlushResume emits it as
//
//	<script type="module" src="URL"></script>
//
// immediately AFTER the resume payload script and BEFORE the trailers, which
// is the order a stock Marko app boots in: the payload defines M._.r and calls
// M._.w() synchronously, then the deferred module executes init() against the
// state that is already there.
//
// Generated code calls this only for PAGE templates that actually compile to
// non-empty dom output; a page with no reactivity ships no bundle.
//
// Resume support; generated code only; not stable.
func (w *Writer) ClientBundle(url string) {
	w.resume().clientBundle = url
}

func newResumeState() *resumeState {
	return &resumeState{
		runtimeID:   DefaultRuntimeID,
		renderID:    DefaultRenderID,
		writeScopes: map[int]*ScopeState{},
		nextScopeID: 1,
		ser:         newSerializer(),
	}
}

// SetResumeIDs overrides the runtime and render ids used for the bootstrap,
// the runtime prefix and the marker comments. Both must match
// /^[_a-zA-Z][_a-zA-Z0-9]*$/, because they are interpolated into the payload
// unescaped; an invalid id is recorded as a render error.
//
// The defaults ("M" and "_") are what a Marko page template produces, so most
// renders never need this.
//
// Resume support; generated code only; not stable.
func (w *Writer) SetResumeIDs(runtimeID, renderID string) {
	r := w.resume()
	if !idPattern.MatchString(runtimeID) || !idPattern.MatchString(renderID) {
		if r.err == nil {
			r.err = errInvalidResumeID
		}
		return
	}
	r.runtimeID = runtimeID
	r.renderID = renderID
}

// resume lazily creates the resume channel, so a render that never touches
// resumability pays nothing.
func (w *Writer) resume() *resumeState {
	if w.res == nil {
		w.res = newResumeState()
	}
	return w.res
}

// AllocScopeID allocates the next scope id for this render, mirroring the JS
// `_scope_id()` intrinsic (a post-increment of a per-render counter).
//
// Ids are 1-BASED -- 0 is reserved for $global -- and allocation ORDER is part
// of the wire format, because the ids appear in markers, in scope references
// and in the ascending delta encoding of the payload. Generated code must
// therefore call this exactly where the compiled JS calls `_scope_id()`,
// including inside body closures and conditional branches: allocation order is
// render order, not source order.
//
// Resume support; generated code only; not stable.
func (w *Writer) AllocScopeID() int {
	r := w.resume()
	id := r.nextScopeID
	r.nextScopeID++
	return id
}

// PeekScopeID returns the id AllocScopeID would hand out next, without
// consuming it. Mirrors the JS `_peek_scope_id()` intrinsic, which the
// compiler uses to name a child scope that a callee is about to allocate.
//
// Resume support; generated code only; not stable.
func (w *Writer) PeekScopeID() int {
	return w.resume().nextScopeID
}

// TouchScope records that a scope EXISTS without giving it any state, mirroring
// the JS `_existing_scope(id)` intrinsic (`writeScope(id, {})`).
//
// A scope with no properties is skipped by the serializer, so on its own this
// writes nothing to the payload -- but a later AddScope for the same id merges
// into it, and it marks the render as needing the resume runtime.
//
// Resume support; generated code only; not stable.
func (w *Writer) TouchScope(id int) ScopeRef {
	w.AddScope(id, NewScopeState())
	return ScopeRef{ID: id}
}

// AddScope records serialized state for a scope id. Repeated calls for the
// same id MERGE their properties rather than replacing them, matching the JS
// writer.
//
// Scope id 0 is $global; ids for template scopes start at 1.
//
// Resume support; generated code only; not stable.
func (w *Writer) AddScope(id int, state *ScopeState) {
	r := w.resume()
	r.needsRuntime = true
	if existing, ok := r.writeScopes[id]; ok {
		existing.merge(state)
		return
	}
	merged := NewScopeState()
	merged.merge(state)
	r.writeScopes[id] = merged
}

// AddScript records a script (effect) entry binding a registry id to a scope
// id. Entries are emitted in CALL ORDER, and consecutive entries sharing a
// registry id collapse: AddScript(1,"R1"), AddScript(2,"R1"), AddScript(3,"R2")
// produces "R1 1 2 R2 3".
//
// The registry id must be non-numeric -- the decoder distinguishes the two
// token kinds by testing for a non-digit.
//
// Resume support; generated code only; not stable.
func (w *Writer) AddScript(registryID string, scopeID int) {
	r := w.resume()
	r.needsRuntime = true
	r.hasEffect = true
	if r.lastEffect == registryID {
		r.effects += " " + strconv.Itoa(scopeID)
		return
	}
	r.lastEffect = registryID
	entry := registryID + " " + strconv.Itoa(scopeID)
	if r.effects == "" {
		r.effects = entry
	} else {
		r.effects += " " + entry
	}
}

// Marker writes a resume marker comment into the HTML stream. Markers are
// placed IMMEDIATELY AFTER the node they name.
//
// The comment is `<!--{runtimeID}{renderID}{op}{scopeID} {accessor}-->`; the
// bootstrap's TreeWalker matches the id prefix and records the node under the
// accessor.
//
// Resume support; generated code only; not stable.
func (w *Writer) Marker(op string, scopeID int, accessor string) {
	r := w.resume()
	r.needsRuntime = true
	w.HTML("<!--" + r.runtimeID + r.renderID + op + strconv.Itoa(scopeID) + " " + accessor + "-->")
}

// Trailer buffers markup that must be written AFTER the resume payload
// script -- in practice the closing `</body></html>` of an <html> tag.
//
// Marko accumulates this separately (`_trailers`) precisely so the payload
// script can be spliced in ahead of it. Generated code should route its final
// closing-tag write here whenever the template has an <html> root; everything
// else keeps using HTML.
//
// Resume support; generated code only; not stable.
func (w *Writer) Trailer(s string) {
	w.trailers.WriteString(s)
}

// BeginTemplate marks the start of one template's render. Generated code emits
// it as the first statement of every render function, paired with FlushResume
// as the last.
//
// It exists because "which template flushes the resume payload?" is a RUNTIME
// question, not a compile-time one. Marko's compiler marks every non-embedded
// template `page` (the third argument to `_template`), so that flag cannot tell
// a page apart from a tag it calls -- in JS the distinction is simply which
// template `.render()` was invoked on. The Go analogue is depth: the outermost
// active render is the document, and it is the one that flushes.
//
// Resume support; generated code only; not stable.
func (w *Writer) BeginTemplate() {
	w.renderDepth++
}

// FlushResume closes one template's render. On the OUTERMOST call it writes the
// resume payload script, then the client bundle script, then any buffered
// trailer markup, completing the document; a nested call only decrements the
// depth and does nothing else, so a tag can never emit a payload mid-document.
//
// Generated code emits it as the last statement of every render function,
// paired with BeginTemplate as the first:
//
//	func Page(w *runtime.Writer, input PageInput) {
//	    w.BeginTemplate()
//	    w.HTML("<html><body>...")
//	    w.Marker(runtime.OpResume, 1, "a")
//	    w.AddScope(1, runtime.ScopeStateOf("c", 0))
//	    w.AddScript("WDGvDBz", 1)
//	    w.Trailer("</body></html>")
//	    w.FlushResume()
//	}
//
// A render with no resume state writes nothing at all -- not even an empty
// <script> -- so calling it unconditionally is safe and costs nothing.
//
// It returns the first error hit while serializing (an unsupported value type
// or a cycle). On error nothing is written, so the caller can surface a
// failure rather than shipping a corrupt payload.
//
// Resume support; generated code only; not stable.
func (w *Writer) FlushResume() error {
	if w.renderDepth > 0 {
		w.renderDepth--
		if w.renderDepth > 0 {
			return nil
		}
	}
	if w.res == nil {
		w.flushTrailers()
		return nil
	}
	r := w.res
	if r.err != nil {
		return r.err
	}

	scripts, err := r.buildScripts()
	if err != nil {
		return err
	}
	if scripts != "" {
		w.sb.WriteString("<script>")
		w.sb.WriteString(scripts)
		w.sb.WriteString("</script>")
	}
	// The browser bundle loads AFTER the payload: init() needs M._.r to
	// already exist. Emitted only when a bundle was named -- a page with no
	// client code, and every byte-oracle comparison, sees nothing here.
	if r.clientBundle != "" {
		w.sb.WriteString(`<script type="module" src="`)
		w.sb.WriteString(EscapeAttr(r.clientBundle))
		w.sb.WriteString(`"></script>`)
	}
	w.flushTrailers()
	return nil
}

// EndTemplate is FlushResume for generated code, which has no way to return an
// error: render functions are `func(*Writer, Input)`, deliberately, so a
// template call reads like a statement.
//
// On a serialization failure it PANICS, naming the template. That is the loud
// end of the spectrum on purpose: the alternative -- swallowing the error --
// ships a page whose HTML looks right and whose interactivity is silently dead,
// which is the worst possible failure mode for a resumability feature. A
// panic surfaces in the http handler's recover, in tests, and in dev, and the
// only way to reach one is a value type the payload cannot express (an
// unsupported Go type in `<let>` state, or a cycle), which is a template bug.
//
// OPEN QUESTION for the error-model review: render functions returning an
// error, or an error accumulated on the Writer and surfaced by String(), would
// both be less violent. Neither fits the current `func(*Writer, Input)` shape,
// so this is deliberately a placeholder with a loud failure rather than a
// quiet one. See the Phase C report.
//
// Resume support; generated code only; not stable.
func (w *Writer) EndTemplate(templateName string) {
	if err := w.FlushResume(); err != nil {
		panic("marko-go: rendering " + templateName + ": " + err.Error())
	}
}

func (w *Writer) flushTrailers() {
	if w.trailers.Len() > 0 {
		w.sb.WriteString(w.trailers.String())
		w.trailers.Reset()
	}
}

// buildScripts assembles the `;`-joined script body: the bootstrap prelude,
// the resume array, and the kick-off call.
func (r *resumeState) buildScripts() (string, error) {
	flushes := make([]scopeFlush, 0, len(r.writeScopes))
	for _, id := range sortedScopeIDs(r.writeScopes) {
		state := r.writeScopes[id]
		// A scope with no own properties is skipped entirely.
		if state.Len() == 0 {
			continue
		}
		flushes = append(flushes, scopeFlush{id: id, state: state})
	}

	resumes, err := r.ser.writeScopesRoot(flushes)
	if err != nil {
		return "", err
	}

	// The effects string is ALWAYS the last element of the resume array,
	// regardless of the order scopes and scripts were recorded in.
	if r.hasEffect {
		if resumes != "" {
			resumes += ","
		}
		resumes += `"` + r.effects + `"`
	}

	if resumes == "" && !r.needsRuntime {
		return "", nil
	}
	if resumes == "" {
		// Markers were emitted but nothing serialized: no script at all.
		return "", nil
	}

	var parts []string
	parts = append(parts, bootstrapRuntimeCode+`("`+r.runtimeID+`")("`+r.renderID+`")`)
	prefix := r.runtimeID + "." + r.renderID
	parts = append(parts, prefix+".r=["+resumes+"]")
	// The kick-off is NOT unconditional: it is emitted only when this flush
	// produced an effects string. A payload of markers and scope data alone
	// does not walk.
	if r.hasEffect {
		parts = append(parts, prefix+".w()")
	}
	return strings.Join(parts, ";"), nil
}

File: packages/marko-go/src/resume.mjs
/**
 * Resumability translators (FR12 wave 1, Phase C).
 *
 * The intrinsics here used to live in `RESUME_ONLY` -- dropped wholesale,
 * because go-marko rendered pages that never hydrated. They now emit real Go,
 * so a generated page ships the same resume payload Marko's own `html`
 * renderer does, byte for byte.
 *
 * ## The model
 *
 * A resume payload has three moving parts, and the compiled JS threads them
 * through intrinsics that all key off ONE thing: the scope id.
 *
 *   scope ids   `_scope_id()` post-increments a per-RENDER counter, 1-based.
 *               The ids are wire-visible (markers, `_(id)` refs, and the
 *               payload's ascending delta encoding), so ALLOCATION ORDER is
 *               part of the contract. It is RENDER order, not source order:
 *               a body closure allocates when the callee renders it, not
 *               where it is written. That is why these are runtime calls
 *               (`w.AllocScopeID()`) and not transpile-time constants -- a
 *               transpile-time allocator cannot see that a `<@head>` body
 *               renders before the page body, or that a tag renders its body
 *               in the middle of its own markup.
 *
 *   markers     `_el_resume(scopeId, accessor, guard?)` writes
 *               `<!--M_*<id> <accessor>-->` immediately after the node it
 *               names -> `w.Marker(runtime.OpResume, id, acc)`.
 *
 *   payload     `_scope(id, {...})` accumulates serialized state and
 *               `_script(id, "regId")` accumulates client-effect entries;
 *               both land in the Writer's resume channel and flush as one
 *               `<script>` at the end of the page.
 *
 * ## Guards (contract sec 10) -- and why they are constants here
 *
 * Marko gates serialization with a caller-deposited "serialize reason":
 * `_set_serialize_reason(r)` deposits, `_scope_reason()` consume-once reads,
 * and `_serialize_if(r, key)` / `_serialize_guard(r, key)` test bit `key+1`.
 *
 * For a PAGE render the whole chain is rooted at zero and stays there:
 *
 *   - Nothing deposits a reason before the page root runs, so the root's
 *     `_scope_reason()` is `undefined` and every guard derived from it is 0.
 *   - A template deposits for its callee only via values it derived from its
 *     OWN reason (`_set_serialize_reason({1: $sg__input_class, ...})`), so a
 *     zero reason propagates as zeros.
 *   - The runtime's only other depositor is `_dynamic_tag`, which deposits
 *     `shouldResume && input !== undefined ? 1 : 0` where
 *     `shouldResume = serializeReason !== 0` -- and the compiler always
 *     passes a `_serialize_guard(...)` result there, i.e. 0.
 *
 * So every `_serialize_guard(...)` is 0 and every `_serialize_if(...)` is
 * falsy for the entire wave-1 surface. That is verified against the real
 * runtime by tracing every fixture page (see the trace evidence in the Phase C
 * report), and it is what lets these translate to Go CONSTANTS: a guarded
 * `_el_resume` emits no marker, a `(_serialize_if(...)) && _scope(...)`
 * statement is dropped, and an unguarded call emits normally.
 *
 * This is exactly the wave-1 relief the wire contract sanctions, widened from
 * "page-root templates" to "the whole tree of a page render" with the
 * propagation argument above. The moment go-marko renders an EMBEDDED template
 * (contract sec 15.5) or a caller deposits a non-zero reason, this becomes
 * wrong -- see the report's wave-2 issues.
 */

import * as t from "@babel/types";
import { UnsupportedError } from "./errors.mjs";
import { goStringLiteral } from "./names.mjs";
// Cyclic with expr.mjs (it imports the translators here, this needs its
// optional-field helper). Fine under ESM live bindings as long as neither side
// touches the other at module-evaluation time -- only inside function bodies.
import { absentIfZero } from "./expr.mjs";

// ---------------------------------------------------------------------------
// Guard evaluation
// ---------------------------------------------------------------------------

/**
 * Statically evaluate a serialize guard/reason expression to its wave-1 value.
 *
 * Returns `0` for anything derived from a serialize reason (see the module
 * header for why that is always the answer here), `undefined` when the
 * expression is not a guard at all.
 *
 * Recognized shapes, all of which the compiler emits:
 *
 *   _serialize_guard($reason, 3)      direct call
 *   _serialize_if($reason, 3)         direct call
 *   $sg__input_href                   a const bound to one of the above
 *   ($sg__x) || ($sg__y)              a union of guards (`_if`'s branch guard)
 *   0 / 1                             a literal the compiler inlined
 */
export function evalGuard(ctx, node) {
  if (node === undefined || node === null) return undefined;
  if (t.isNumericLiteral(node)) return node.value;
  if (t.isIdentifier(node)) {
    return ctx.resume.guardConsts.has(node.name)
      ? ctx.resume.guardConsts.get(node.name)
      : undefined;
  }
  const intrinsic = calleeIntrinsicName(ctx, node);
  if (intrinsic === "_serialize_guard") return 0;
  if (intrinsic === "_serialize_if") return 0;
  if (intrinsic === "_scope_reason") return 0;
  if (t.isLogicalExpression(node) && (node.operator === "||" || node.operator === "&&")) {
    const l = evalGuard(ctx, node.left);
    const r = evalGuard(ctx, node.right);
    if (l === undefined || r === undefined) return undefined;
    // Both sides are 0 in wave 1, so either operator yields 0. Written out
    // rather than hardcoded so a future non-zero reason surfaces here.
    return node.operator === "||" ? l || r : l && r;
  }
  return undefined;
}

/**
 * A guard ARGUMENT decides whether a marker/scope is emitted at all. Absent
 * (the compiler omitted the parameter) means "always emit"; present means
 * "emit iff truthy", which in wave 1 is never.
 */
export function guardAllowsEmit(ctx, node) {
  if (node === undefined) return true;
  const value = evalGuard(ctx, node);
  if (value === undefined) {
    throw new UnsupportedError(
      "unrecognized serialize-guard expression -- the resume translators only understand guards derived from _scope_reason()",
      node,
    );
  }
  return Boolean(value);
}

/**
 * Is this a `(guard) && <expr>` whose guard is statically off? Those statements
 * (`(_serialize_if(...)) && _scope(...)`) drop entirely.
 */
export function isDisabledGuardedStatement(ctx, node) {
  if (!t.isLogicalExpression(node) || node.operator !== "&&") return false;
  if (!isGuardExpression(ctx, node.left)) return false;
  const value = evalGuard(ctx, node.left);
  return value !== undefined && !value;
}

/**
 * Record `const $sg__x = _serialize_guard(...)` style bindings so a later
 * reference to `$sg__x` evaluates to the same constant. Returns true when the
 * declarator was a guard binding (and therefore emits no Go).
 *
 * Only an initializer that is ITSELF a guard construct counts. A bare literal
 * does NOT -- `let count = 0` is a `<let>` state variable, and treating its `0`
 * as a guard would silently delete the template's state.
 */
export function recordGuardConst(ctx, decl) {
  if (!t.isIdentifier(decl.id)) return false;
  if (!isGuardExpression(ctx, decl.init)) return false;
  const value = evalGuard(ctx, decl.init);
  if (value === undefined) return false;
  ctx.resume.guardConsts.set(decl.id.name, value);
  return true;
}

/**
 * Is this expression rooted in the serialize-guard machinery, as opposed to
 * something that merely happens to evaluate to a number? Bare literals are
 * meaningful as guards only in guard POSITION (an argument slot), never as the
 * thing that decides an expression is a guard.
 */
function isGuardExpression(ctx, node) {
  if (!node) return false;
  const intrinsic = calleeIntrinsicName(ctx, node);
  if (
    intrinsic === "_serialize_guard" ||
    intrinsic === "_serialize_if" ||
    intrinsic === "_scope_reason"
  ) {
    return true;
  }
  if (t.isIdentifier(node)) return ctx.resume.guardConsts.has(node.name);
  if (t.isLogicalExpression(node)) {
    return isGuardExpression(ctx, node.left) || isGuardExpression(ctx, node.right);
  }
  return false;
}

// ---------------------------------------------------------------------------
// Scope ids
// ---------------------------------------------------------------------------

/**
 * `const $scopeN_id = _scope_id()` -> `scopeN := w.AllocScopeID()`.
 *
 * The Go name is derived from the JS binding rather than invented, so a later
 * `_scope($scopeN_id, ...)` in the same body resolves by plain identifier
 * lookup. Names are sanitized (`$` is not a Go identifier character) and
 * recorded on ctx so `scopeIdExpr` can find them.
 */
export function translateScopeIdDecl(ctx, decl) {
  const goName = goScopeVarName(decl.id.name);
  ctx.resume.scopeVars.set(decl.id.name, goName);
  ctx.resume.used = true;
  return `${goName} := w.AllocScopeID()`;
}

/** `const $childScope = _peek_scope_id()` -> `x := w.PeekScopeID()`. */
export function translatePeekScopeIdDecl(ctx, decl) {
  const goName = goScopeVarName(decl.id.name);
  ctx.resume.scopeVars.set(decl.id.name, goName);
  ctx.resume.used = true;
  return `${goName} := w.PeekScopeID()`;
}

/**
 * Drop the binding from `X := w.AllocScopeID()` when nothing in the body ever
 * reads `X`, leaving the bare call.
 *
 * The CALL cannot be dropped -- it advances the render's scope-id counter, and
 * every later id (and therefore every marker, `_(id)` reference and payload
 * delta) shifts if it is missing. Only the Go variable is unnecessary, and Go
 * rejects an unused one outright, so this is required rather than cosmetic.
 * It happens here, at assembly, because whether a scope id is read is only
 * known once the entire body is translated: a template can allocate a scope
 * purely for ordering (every icon tag does) and never mention it again.
 */
export function dropUnusedScopeVars(ctx, bodyGo) {
  if (ctx.resume.scopeVars.size === 0) return bodyGo;
  const joined = bodyGo.join("\n");
  const unused = new Set();
  for (const goName of ctx.resume.scopeVars.values()) {
    const uses = joined.match(new RegExp(`\\b${goName}\\b`, "g"));
    if (uses && uses.length === 1) unused.add(goName);
  }
  if (unused.size === 0) return bodyGo;
  // Entries can themselves be MULTI-LINE (a body closure is one string), so
  // the rewrite walks lines inside each entry rather than whole entries.
  const rewriteLine = (line) => {
    const m = /^(\s*)([A-Za-z_][A-Za-z0-9_]*) := (w\.(?:Alloc|Peek)ScopeID\(\))$/.exec(line);
    return m && unused.has(m[2]) ? `${m[1]}${m[3]}` : line;
  };
  return bodyGo.map((entry) => entry.split("\n").map(rewriteLine).join("\n"));
}

/** `$scope0_id` / `$childScope` -> the Go int variable holding that scope id. */
export function scopeIdExpr(ctx, node) {
  if (t.isNumericLiteral(node)) return String(node.value);
  if (t.isIdentifier(node)) {
    const goName = ctx.resume.scopeVars.get(node.name);
    if (goName) return goName;
  }
  throw new UnsupportedError(
    "expected a scope id bound by _scope_id()/_peek_scope_id()",
    node,
  );
}

/**
 * Marko names scope-id bindings `$scope0_id`, `$scope1_id`, `$childScope`.
 * `$` is not legal in a Go identifier, so it is dropped; the rest is already a
 * valid Go name and is left alone so the generated code stays readable next to
 * the compiled JS it mirrors.
 */
function goScopeVarName(jsName) {
  const cleaned = jsName.replace(/\$/g, "");
  return /^[A-Za-z_]/.test(cleaned) ? cleaned : `s${cleaned}`;
}

// ---------------------------------------------------------------------------
// Markers, scripts, scopes
// ---------------------------------------------------------------------------

/**
 * `_el_resume(scopeId, "acc", guard?)` -> `runtime.Marker(...)` as an HTML
 * fragment expression, or `""` when the guard suppresses it.
 *
 * It appears inside an `_html` template literal, so it must be an EXPRESSION;
 * the Writer method is `w.Marker(...)`, which writes directly, so the emitting
 * side (translateHtmlCall) special-cases this one to a statement. See
 * `elResumeStatement`.
 */
export function elResumeStatement(ctx, node) {
  if (!guardAllowsEmit(ctx, node.arguments[2])) return null;
  const [scopeArg, accessorArg] = node.arguments;
  if (!t.isStringLiteral(accessorArg)) {
    throw new UnsupportedError("_el_resume accessor is not a string literal", node);
  }
  ctx.resume.used = true;
  return `w.Marker(runtime.OpResume, ${scopeIdExpr(ctx, scopeArg)}, ${goStringLiteral(accessorArg.value)})`;
}

/** `_script(scopeId, "regId")` -> `w.AddScript("regId", scopeId)`. */
export function translateScript(ctx, node) {
  const [scopeArg, regArg] = node.arguments;
  if (!t.isStringLiteral(regArg)) {
    throw new UnsupportedError("_script registry id is not a string literal", node);
  }
  ctx.resume.used = true;
  ctx.resume.registryIds.add(regArg.value);
  return `w.AddScript(${goStringLiteral(regArg.value)}, ${scopeIdExpr(ctx, scopeArg)})`;
}

/**
 * `_scope(scopeId, { k: v, ... })` -> `w.AddScope(id, runtime.ScopeStateOf(...))`.
 *
 * Key ORDER is preserved as written: ScopeStateOf records insertion order and
 * the serializer applies JS enumeration order (integer-like keys first) on top,
 * so the Go side reproduces `for (const key in obj)` exactly.
 *
 * A property whose value is a statically-off guard (`k: (_serialize_if(...)) &&
 * expr`) is DROPPED rather than emitted as `false`: that is what JS produces
 * too, since `_serialize_if` returns `undefined` and the serializer omits an
 * undefined property.
 */
export function translateScope(ctx, node, translateValue) {
  const [scopeArg, objArg] = node.arguments;
  if (!t.isObjectExpression(objArg)) {
    throw new UnsupportedError("_scope called with a non-literal state object", node);
  }
  ctx.resume.used = true;
  const pairs = [];
  for (const prop of objArg.properties) {
    if (!t.isObjectProperty(prop) || prop.computed) {
      throw new UnsupportedError("unsupported property in _scope state", prop);
    }
    const key = t.isIdentifier(prop.key)
      ? prop.key.name
      : t.isStringLiteral(prop.key)
        ? prop.key.value
        : t.isNumericLiteral(prop.key)
          ? String(prop.key.value)
          : null;
    if (key === null) throw new UnsupportedError("unsupported _scope key", prop.key);
    // `k: (guard) && value` with the guard off -> the property never exists.
    if (isDisabledGuardedStatement(ctx, prop.value)) continue;
    pairs.push(`${goStringLiteral(key)}, ${translateScopeValue(ctx, prop.value, translateValue)}`);
  }
  const id = scopeIdExpr(ctx, scopeArg);
  if (pairs.length === 0) {
    // `_scope(id, {})` still marks the scope as existing (and the render as
    // needing the runtime) even though an empty partial serializes to nothing.
    return `w.TouchScope(${id})`;
  }
  return `w.AddScope(${id}, runtime.ScopeStateOf(${pairs.join(", ")}))`;
}

/**
 * One `_scope` property value. Scope-valued intrinsics become `runtime.Scope`
 * / `w.TouchScope` (both `ScopeRef`, which the serializer writes as `_(id)`);
 * anything else falls through to the ordinary expression translator.
 */
function translateScopeValue(ctx, node, translateValue) {
  const intrinsic = calleeIntrinsicName(ctx, node);
  if (intrinsic === "_scope_with_id") {
    return `runtime.Scope(${scopeIdExpr(ctx, node.arguments[0])})`;
  }
  if (intrinsic === "_existing_scope") {
    // `_existing_scope(id)` is `writeScope(id, {})` -- it BOTH registers the
    // scope and evaluates to a reference to it.
    return `w.TouchScope(${scopeIdExpr(ctx, node.arguments[0])})`;
  }
  // An optional input field that was never passed must serialize as
  // `undefined` -- dropped from the object -- not as its Go zero value.
  return absentIfZero(ctx, node, translateValue(ctx, node));
}

/** `_trailers("</body></html>")` -> `w.Trailer("...")`. */
export function translateTrailers(ctx, node) {
  const [arg] = node.arguments;
  if (!t.isStringLiteral(arg)) {
    throw new UnsupportedError("_trailers called with a non-literal argument", node);
  }
  ctx.resume.used = true;
  return arg.value === "" ? [] : [`w.Trailer(${goStringLiteral(arg.value)})`];
}

/**
 * `_resume_branch(scopeId)` records the CLOSEST enclosing resumed branch on a
 * scope (`writeScope(id, {ClosestBranchId: branchId})`), and only when the
 * render is actually inside one -- which requires `<await>`, out-of-order
 * streaming, or a `_dynamic_tag` rendering a resumable component branch.
 *
 * None of that exists in the wave-1 subset (no async, no `<await>`, guards
 * uniformly off so `withBranchId` is never entered), so the call is a
 * guaranteed no-op and emits nothing. Revisit with the control-flow /
 * streaming waves.
 */
export function translateResumeBranch() {
  return [];
}

// ---------------------------------------------------------------------------

/** Local helper mirroring expr.mjs's calleeIntrinsic without importing it. */
function calleeIntrinsicName(ctx, node) {
  if (!t.isCallExpression(node) || !t.isIdentifier(node.callee)) return undefined;
  return ctx.localToIntrinsic.get(node.callee.name);
}
